src/engine/controls/cuecontrol.cpp

```diff
@@ -765,7 +896,95 @@ void CueControl::hotcueGotoAndPlay(HotcueControl* pControl, double v) {
     }
 }
 
-void CueControl::hotcueActivate(HotcueControl* pControl, double v) {
+void CueControl::hotcueGotoAndLoop(HotcueControl* pControl, double v) {
+    if (!v) {
+        return;
+    }
+
+    QMutexLocker lock(&m_mutex);
+    if (!m_pLoadedTrack) {
+        return;
+    }
+
+    CuePointer pCue(pControl->getCue());
+
+    // Need to unlock before emitting any signals to prevent deadlock.
+    lock.unlock();
+
+    if (!pCue) {
+        return;
+    }
+
+    double startPosition = pCue->getPosition();
+    if (startPosition == Cue::kNoPosition) {
+        return;
+    }
+
+    if (pCue->getType() == mixxx::CueType::Loop) {
+        hotcueGoto(pControl, v);
+        setCurrentSavedLoopControl(pControl);
+    } else if (pCue->getType() == mixxx::CueType::HotCue) {
+        hotcueGoto(pControl, v);
+        setBeatLoop(pCue->getPosition(), true);
+    } else {
+        return;
+    }
+
+    if (!isPlayingByPlayButton()) {
+        // cueGoto is processed asynchronously.
+        // avoid a wrong cue set if seek by cueGoto is still pending
+        m_bPreviewing = false;
+        m_iCurrentlyPreviewingHotcues = 0;
+        // don't move the cue point to the hot cue point in DENON mode
+        m_bypassCueSetByPlay = true;
+        m_pPlay->set(1.0);
+    }
+}
+
+void CueControl::hotcueLoopToggle(HotcueControl* pControl, double v) {
```

Comment:
I use this CO for catching loops (in contrast to jumping to the loop's start position). For a controller with a dedicated loop mode (e.g. Serato-style controllers), you can map the regular activate to pad press and this CO to SHIFT+pad press (or vice versa).

Reply:
I think you misunderstood me. You commented on the `hotcue_N_loop_toggle` CO. We can't use the `reloop_toggle` or `loop_enabled` here, because without this CO there is no way to enable the saved loop without jumping.

Regarding `gotoandloop`: I don't see what this has to do with previewing. `gotoandloop` is used by `activate` for saved loops, just like it uses `gotoandplay` for regular hotcues.

Reply:
My hope is to de-clutter all that if we have a plan how to issue serato and tracktor type controllers in combination with the GUI.

Reply:
If the GUI should follow the controller strategy, we need controls that have a changing behavior controlled by preferences and not fixed features like proposed here.

Reply:
But I am unsure. Alternative is to allow always missed assignment and hope that the user does not mass it up on the serato controller.